File: solutions/1610/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Wed Dec 15 16:23:47 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <cmath> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int visiblePoints(vector<vector<int>>& points, int angle, vector<int>& location) {
        int result = 0;
        vector<double> ptsAng;
        ptsAng.reserve(2*points.size());
        // convert origin to person
        int ptsOriNum = 0;
        for (auto& p : points) {
            if (p[0] == location[0] && p[1] == location[1]) ptsOriNum++;
            else {
                p[0] -= location[0];
                p[1] -= location[1];
                ptsAng.push_back(atan2(p[1], p[0]));
            }
        }
        // sort the angles
        sort(ptsAng.begin(), ptsAng.end());
        const int len = ptsAng.size();
        for (int i = 0; i < len; ++i) ptsAng.push_back(ptsAng[i] + 2.0*M_PI);
        // sliding window
        int l = 0;
        double fov = M_PI * angle / 180;
        for (int r = 0; r < ptsAng.size(); ++r) {
            while (ptsAng[r] - ptsAng[l] > fov) l++;
            result = max(result, r - l + 1);
        }
        return result + ptsOriNum;
    }
};

TEST(Test1610, SimpleTest) {
    vector<vector<int>> points{{2,1},{2,2},{3,3}};
    const double angle = 90; 
    vector<int> location{
        1,1
    };
    Solution s; 
    EXPECT_EQ(s.visiblePoints(points, angle, location), 3);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/1610/main.py
from typing import List 

import math 
class Solution:
    def visiblePoints(self, points: List[List[int]], angle: int, location: List[int]) -> int:
        pts_origin, result = 0, 0
        pts_angles = []
        # convert the origin to person location 
        for p in points:
            p[0], p[1] = p[0] - location[0], p[1] - location[1]
        # check how many points are at origin 
        for p in points: 
            if p[0] == 0 and p[1] == 0: 
                pts_origin += 1 
            else: 
                a = math.atan2(p[1], p[0])
                pts_angles.append(a)
        # sort the point angles 
        pts_angles = sorted(pts_angles)
        length = len(pts_angles)
        for i in range(length): 
            pts_angles.append(pts_angles[i] + 2.0 * math.pi)
        # sliding window 
        l = 0
        fov = math.pi * angle / 180 
        # NOTE, time complexity in for+while step is O(n), not O(n^2)
        # since l traverse takes O(n) in total
        for r in range(len(pts_angles)): 
            while pts_angles[r] - pts_angles[l] > fov: 
                l += 1 
            result = max(result, r - l + 1)
        return result + pts_origin
        
if __name__ == "__main__": 
    points = [[2,1],[2,2],[3,3]]
    angle = 90
    location = [1,1]
    s = Solution()
    result = s.visiblePoints(points, angle, location)
    assert result == 3
